src/util/indexing.jl

```diff
@@ -0,0 +1,263 @@
+    names!(x, convert(Vector{Symbol}, names))
+end
+
+
+function rename!(x::Indexer, names)
+    for (from, to) in names
+        if !haskey(x, from)
+            throw(ArgumentError("Cannot rename $from to $to, $from does not exist in the Indexer."))
+        end
+        if haskey(x, to)
+            throw(ArgumentError("Cannot rename $from to $to, $to already exists in the Indexer."))
+        end
+        # Change the name array
+        nameidx = findfirst(x.names, from)
+        x.names[nameidx] = to
+        # Change the dictionary
+        x.lookup[to] = val = pop!(x.lookup, from)
+    end
+    return x
+end
+
+
+"""
+Rename a set of names in an Indexer, to a new set of names.
+
+This will rename such that the name in from[1] will be renamed to to[1],
+from[2] will be renamed to to[2] and so on.
+"""
+function rename!(x::Indexer, from::Vector{Symbol}, to::Vector{Symbol})
+    return rename!(x, zip(from, to))
+end
+
+
+"""
+Rename a set of names in an Indexer, to a new set of names.
+
+This will rename such that the name in from[1] will be renamed to to[1],
+from[2] will be renamed to to[2] and so on.
+"""
+function rename!{S <: AbstractString}(x::Indexer, from::Vector{S}, to::Vector{S})
+    return rename!(x, convert(Vector{Symbol}, from), convert(Vector{Symbol}, to))
+end
+
+
+"Rename a single name (from) in Indexer x, to a new name (to)."
+rename!(x::Indexer, from::Symbol, to::Symbol) = rename!(x, ((from, to),))
+
+"Rename a single name (from) in Indexer x, to a new name (to)."
+function rename!{S <: AbstractString}(x::Indexer, from::S, to::S)
+    return rename!(x, ((convert(Symbol, from), convert(Symbol, to)),))
+end
+
+
+"""
+Rename a indexer according to some function, f, that accepts one argument.
+
+Each name in the Indexer x, will be input to, f, and the output will be
+what that name will be renamed to.
+
+Therefore the function, f, will need to return Symbols or Strings.
+"""
+rename!(x::Indexer, f::Function) = rename!(x, [(x,f(x)) for x in x.names])
+rename(x::Indexer, args...) = rename!(copy(x), args...)
+
+
+# Indexing into an index mapping single names to single numbers
+
+## Indexing with single name
+Base.getindex(x::Indexer, idx::Symbol) = x.lookup[idx]
+Base.getindex{S <: AbstractString}(x::Indexer, idx::S) = x.lookup[convert(Symbol, idx)]
+
+## Indexing with multiple names.
+Base.getindex(x::Indexer, idx::Vector{Symbol}) = [x.lookup[i] for i in idx]
+function Base.getindex{S <: AbstractString}(x::Indexer, idx::Vector{S})
+    return [x.lookup[convert(Symbol, i)] for i in idx]
+end
+
+## Indexing with a single integer
+function Base.getindex{I <: Integer}(x::Indexer, idx::I)
+    n = x.names[idx]
+    return x[n]
+end
+
+## Indexing with a vector of intergers
+function Base.getindex{I <: Integer}(x::Indexer, idx::Vector{I})
+    ns = x.names[idx]
+    return x[ns]
+end
+
+## Indexing with a range of integers
+function Base.getindex{I <: Integer}(x::Indexer, idx::UnitRange{I})
```

Comment:
I think `Base.getindex{I <: Integer}(x::Indexer, idx::Vector{I})` and `Base.getindex{I <: Integer}(x::Indexer, idx::UnitRange{I})` can be merged in to a single definition as `Base.getindex{I<:Integer}(x::Indexer, keys::AbstractVector{I})`.